treewatch.1
.\"  -*- nroff -*-
.\"
.\" treewatch.1
.\"
.\" Author: Roger Light <roger@atchoo.org>
.\" Copyright (C) Roger Light <roger@atchoo.org>, UK
.\"               All rights reserved
.\"
.\" Created:
.\"
.\" $$
.\"
.\" .Th treewatch 1 local
.Dd 1 March, 2008
.Dt TREEWATCH 1
.Os GNU/Linux
.Sh NAME
.Nm treewatch
.Nd watch directories and execute program on file changes
.Sh SYNOPSIS
.Nm
.Bk -words
.Op Fl h
.Nm
.Bk -words
.Op Fl a
.Op Fl c Ar command
.Op Fl d Ar directory
.Op Fl f Ar file
.Op Fl o Ar 'some options'
.Op Fl v
.Op Fl w
.Sh DESCRIPTION
.Nm
is a program to watch a set of directories and execute a program when there is a change in the files within the directories. It was originally conceived as a way of watching directories of source code (.c, .cpp and .h files) and executing make when the files changed and this is reflected in its default mode of operation which is to watch the current directory for changes in .c, .cpp and .h files and execute /usr/bin/make on changes.
.Pp
Any number of directories can be watched, as well as any number of different file endings. It is also possible to watch for changes on all files within a directory.
.Pp
.Nm
will only terminate if it is unable to execute the child command for any reason. 
.Sh OPTIONS
.Bl -tag -width indent
.It Fl a, Fl -all-files
Watch all files for changes in the directories being watched for rather than just the ones with specified file endings. This option makes any use of --file-ending redundant.
.It Fl c, Fl -command Ar command
Specify the child command to run on file changes. The default action is to run /usr/bin/make.
.It Fl d, Fl -directory Ar directory
Specify a directory to watch. This only watches the specific directory given. Any number of directories can be watched at once. If this option is not used, the default action is to watch the current directory.
.It Fl f, Fl -file-ending Ar file
Specify file endings to watch for changes. Any number of file endings can be watched at once. Although this option strictly works on file endings, specifying complete file names works as well. If not given,
.Nm
watches .c, .cpp and .h files for changes.
.It Fl h, Fl -help
Show treewatch usage.
.It Fl o, Fl -options Ar 'some options'
Give a list of options to pass to the child command. Enclose multiple options with quotes. The string "%f" can be used to obtain the name of the file that changed.
.It Fl v, Fl -verbose
Be more verbose. This is currently limited to printing the return status of the child command (assuming that --no-wait is not given).
.It Fl w, Fl -no-wait
Don't wait for the child command to finish before continuing. The default setting is to wait so that a marker can be printed after the child has exited.
.El
.Sh EXAMPLES
Running 
.Nm
without any options is equivalent to running:
.Pp
.Nm
-c /usr/bin/make -d . -f .c -f .cpp -f .h
.Pp
Watch /tmp for any file changes and echo a message about which file changed:
.Pp
.Nm
-a -c /bin/echo -o "tmp file changed: %f" -d /tmp
.Sh BUGS
Only the first "%f" string is replaced in the options argument.
.Sh AUTHORS
.An Roger Light Aq roger@atchoo.org


ChangeLog
1.1.0 - 20080302
================

* Added man page.
* Added the --all-files option which causes treewatch to watch all files in
  the directories being watched rather than just the specified endings.
* Added the --verbose which causes the child command exit status to be
printed.
* Added filename substitution to the --options parsing so that %f is replaced
with the name of the file which changed.


treewatch.c
/*
 * File: treewatch.c Author: Roger Light
 * Date: 2008/02/24
 * Desc: Watch a list of directories and execute a program on a file change.
 *
 * Portions of this file are taken from the GGZ Gaming Zone and are Copyright
 * Josef Spillner.
 *
 * optlist() and simpleexec() come from grubby/grubby/modules/exec.c The bulk
 * of reconfiguration_handle() and main() come from ggzd/ggzd/control.c
 *
 * Copyright (C) 2008 Roger Light.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program; if not, write to the Free Software Foundation, Inc., 59 Temple
 * Place - Suite 330, Boston, MA 02111-1307, USA
 */

#include <fcntl.h>
#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <unistd.h>

#include <sys/inotify.h>
#include <sys/ioctl.h>

#define INOTIFY_EVENTSIZE sizeof(struct inotify_event)

static struct option options[] = {
	{"all-files", no_argument, 0, 'a'},
	{"command", required_argument, 0, 'c'},
	{"directory", required_argument, 0, 'd'},
	{"file-endings", required_argument, 0, 'f'},
	{"command-options", required_argument, 0, 'o'},
	{"help", no_argument, 0, 'h'},
	{"verbose", no_argument, 0, 'v'},
	{"wait", no_argument, 0, 'w'},
	{0, 0, 0, 0}
};

/* The inotify file descriptor. */
int reconfigure_fd;
/* The child command to run. */
char *command;
/* The child command options. */
char *command_options;
/* List of file ending strings to watch. */
char **fileendings;
/* Number of file endings in the list. */
int fileendings_count;
/* Should we wait for the child command to finish before continuing. */
int waitforcommand;
/* Should we watch all files for changes rather than specific endings? */
int all_files;
/* Be verbose? */
int verbose;


/* Build up an option list for the simpleexec function */
static char *const *optlist(char *cmdline, char *options)
{
	static char **list = NULL;
	int i;
	char *token;
	int listsize;

	/* Free list first if already initialized*/
	if(list)
	{
		i = 0;
		while(list[i])
		{
			free(list[i]);
			i++;
		}
		free(list);
	}

	/* Create list from words in the option string */
	list = (char**)malloc(2 * sizeof(char*));
	list[0] = (char*)malloc(strlen(cmdline) + 1);
	strcpy(list[0], cmdline);
	list[1] = NULL;
	listsize = 2;
	token = strtok(options, " ");
	while(token)
	{
		listsize++;
		list = (char**)realloc(list, listsize * sizeof(char*));
		list[listsize - 2] = (char*)malloc(strlen(token) + 1);
		strcpy(list[listsize - 2], token);
		list[listsize - 1] = NULL;
		token = strtok(NULL, " ");
	}

	return list;
}


/* Just execute a program without any checking */
static void simpleexec(const char *cmdline, const char *options)
{
	char *const *opts = optlist((char*)cmdline, (char*)options);

	execvp(cmdline, opts);
}


static void reconfiguration_handle(void)
{
	int pending;
	char buf[4096];
	struct inotify_event ev;
	int offset;
	int diff;
	char *filename;
	pid_t pid;
	int i;
	int status;
	char *modified_options;
	char *pos;

	offset = 0;

	ioctl(reconfigure_fd, FIONREAD, &pending);
	while(pending > 0) {
		if (pending > (int)sizeof(buf)) pending = sizeof(buf);
		pending = read(reconfigure_fd, buf, pending);

		while(pending > 0) {
			memcpy(&ev, &buf[offset], INOTIFY_EVENTSIZE);
			filename = malloc(ev.len + 1);
			if(!filename){
				fprintf(stderr, "Error: Out of memory\n");
				exit(-1);
			}
			memcpy(filename, &buf[offset + INOTIFY_EVENTSIZE], ev.len);

			diff = INOTIFY_EVENTSIZE + ev.len;
			pending -= diff;
			offset += diff;

			for(i = 0; i < fileendings_count; i++){
				if(all_files || !strncmp(filename + strlen(filename) - strlen(fileendings[i]), 
							fileendings[i], strlen(fileendings[i]))){
					i = fileendings_count;
					pid = fork();
					switch(pid) {
						case -1:
							break;
						case 0:
							if(strstr(command_options, "%f")){
								modified_options = calloc((strlen(command_options) - 2 + strlen(filename) + 1),
											sizeof(char));

								if(!modified_options){
									fprintf(stderr, "Error: Out of memory\n");
									exit(-1);
								}

								sprintf(modified_options, "%s", command_options);
								pos = strstr(modified_options, "%f");
								pos[0] = '\0';

								modified_options = strcat(modified_options, filename);
								modified_options = strcat(modified_options, strstr(command_options, "%f")+2);
								simpleexec(command, modified_options);
							}else{
								simpleexec(command, command_options);
							}
							/* If the fork() fails, exit. */
							exit(-1);
							break;
						default:
							if(waitforcommand){
								waitpid(pid, &status, 0);
								if(verbose){
									printf("--- %s exited with status %d ---\n", command, status);
								}
							}
							break;
					}
				}
			}

			free(filename);
		}
	}
}


int main(int argc, char *argv[])
{
	int status;
	int sopt, optindex;
	/* has the user used a -d argument? If not then use the default directory. */
	int have_directory = 0;

	waitforcommand = 1;
	command = NULL;
	command_options = NULL;
	fileendings_count = 0;
	fileendings = NULL;
	all_files = 0;
	verbose = 0;

	/* Initialise inotify */
	fd_set active_fd_set, read_fd_set;
	reconfigure_fd = inotify_init();
	if(reconfigure_fd <= 0) {
		fprintf(stderr, "Error: inotify initialization failed.\n");
		return 1;
	}

	/* Parse arguments */
	while(1) {
		sopt = getopt_long(argc, argv, "ac:d:f:ho:vw", options, &optindex);
		if(sopt == -1) break;
		switch(sopt) {
			case 'a':
				all_files = 1;
				break;
			case 'c':
				command = optarg;
				break;
			case 'd':
				/* Specify directories to watch */
				inotify_add_watch(reconfigure_fd, optarg, IN_CLOSE_WRITE | IN_DELETE | IN_MOVED_TO);
				have_directory = 1;
				break;
			case 'f':
				fileendings_count++;
				fileendings = realloc(fileendings, sizeof(char *)*fileendings_count);
				fileendings[fileendings_count-1] = strdup(optarg);
				break;
			case 'h':
	 			printf("treewatch version %s (%s)\n\n", VERSION, BUILDDATE),
				printf("Copyright (C) 2008 Roger Light\nhttp://atchoo.org/tools/treewatch/\n\n");
				printf("treewatch comes with ABSOLUTELY NO WARRANTY.  You may distribute treewatch freely\nas described in the COPYING file distributed with this file.\n\n");
				printf("treewatch is a program to watch a directory and execute a program on file changes.\n\n");
				printf("Usage: treewatch -h\n");
				printf("       treewatch [-aw] [-c command] [-d dir] [-f file] [-o \"some options\"]\n\n");
				printf(" -a, --all-files      Watch all files in the directories. Makes any --file-ending \n");
				printf("                      arguments redundant.\n");
				printf(" -c, --command        Specify full path to command to run (default: /usr/bin/make)\n");
				printf(" -d, --directory      Directory to watch. May be specified multiple times.\n");
				printf("                      (default: current directory)\n");
				printf(" -f, --file-ending    File endings to watch. May be specified mutiple times.\n");
				printf("                      (default is all of: .c .cpp .h)\n");
				printf(" -h, --help           Display this help.\n");
				printf(" -o, --options        Options to pass to the command to run.\n");
				printf(" -v, --verbose        Be more verbose (print child exit status).\n");
				printf(" -w, --no-wait        Don't wait for child command to terminate.\n");
				printf("\nSee http://atchoo.org/tools/treewatch/ for updates.\n");
				exit(0);
				break;
			case 'o':
				command_options = optarg;
				break;
			case 'v':
				verbose = 1;
				break;
			case 'w':
				waitforcommand = 0;
				break;
			default:
				exit(0);
				break;
		}
	}

	/* Set up the defaults if the user hasn't supplied the given arguments. */
	if(!command) command = strdup("/usr/bin/make");
	if(!command_options) command_options = strdup("");

	if(!have_directory){
		inotify_add_watch(reconfigure_fd, ".", IN_CLOSE_WRITE | IN_DELETE | IN_MOVED_TO);
	}
	if(!fileendings_count){
		fileendings_count = 3;
		fileendings = malloc(sizeof(char *)*fileendings_count);
		fileendings[0] = strdup(".c");
		fileendings[1] = strdup(".h");
		fileendings[2] = strdup(".cpp");
	}

	FD_ZERO(&active_fd_set);
	FD_SET(reconfigure_fd, &active_fd_set);

	/* Main program loop - broken out of only by user interrupt. */
	while(1){
		read_fd_set = active_fd_set;

		status = select(reconfigure_fd + 1, &read_fd_set, NULL, NULL, NULL);

		if(FD_ISSET(reconfigure_fd, &read_fd_set)) {
			reconfiguration_handle();
			}
	}

	if(command) free(command);

	return 0;
}

